### DISCUSSION TITLE
What's the best way to format dates in structs?

### DISCUSSION THREAD

**Discussion by rgwood** [mitsuhiko/minijinja#290] [Q&A] [answered]
I've got some structs like these that get passed to Minijinja:

```rust
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct Item {
    pub id: i64,
    pub name: String,
    pub timestamp: time::Date,
    ...
}
```

If I reference that `timestamp` field in a template, it comes out formatted rather terribly like `[2023, 32]` (which I assume is the default serialization for `time::Date`). So I need to find a way to format dates in the Minijinja output. Some ideas I've had:

1. Create a new `ItemForTemplate` struct with a `timestamp_for_display: String` field that I populate manually in Rust. This would work but I really don't want to duplicate every struct...
2. Create a `date_format` filter and reference it in the template like `{{ timestamp | date_format }}`
    - This seems like a non-starter because filters can't accept `Date` arguments: `the trait for<'a> minijinja::filters::Filter<_, <_ as FunctionArgs<'a>>::Output> is not implemented for fn item for<'a> fn(&'a time::Date) -> std::string::String {format_date}`
    - I can't implement [`ArgType`](https://docs.rs/minijinja/0.34.0/minijinja/value/trait.ArgType.html) for `time::Date`, but I guess Minijinja could (and put the implementation behind a feature). Or I could wrap `time::Date` in a newtype and implement `ArgType` maybe.
3. Create my own custom serializer that uses `serializing_for_value()` to special-case serialization for Minijinja
    - Downside: dates would have to be formatted the same way everywhere

Am I missing any other approaches? I'd be open to switching to `chrono` if that makes it easier.

**Comment by rgwood** [comment, marked as answer]
Date formatting functionality has been added: https://github.com/mitsuhiko/minijinja/pull/309

**Reply by mitsuhiko** [reply]
I only added `DateTime` so far, so `Date` alone is not supported yet (and neither is `Time`). I'm curious if that is used in practice much.

**Reply by rgwood** [reply]
> I'm curious if that (`Date`) is used in practice much.

I often work on applications where things are associated with a specific date without a time. For example: it's pretty common in finance to track end-of-day stock prices, and by definition there is only one of those per stock per day.

Not sure how representative my experience is :)

**Reply by mitsuhiko** [reply]
Shouldn't be hard too add. Happy to accept a PR :)
